lms/djangoapps/experiments/flags.py

```diff
@@ -162,16 +180,28 @@ def get_bucket(self, course_key=None, track=True):
             if not self._is_enrollment_inside_date_bounds(values, user, course_key):
                 return self._cache_bucket(experiment_name, 0)
 
-        bucket = stable_bucketing_hash_group(experiment_name, self.num_buckets, user.username)
-
-        # Now check if the user is forced into a particular bucket, using our subordinate bucket flags
+        # Determine the user's bucket.
+        # First check if forced into a particular bucket, using our subordinate bucket flags.
+        # If not, calculate their default bucket using a consistent hash function.
         for i, bucket_flag in enumerate(self.bucket_flags):
-            if bucket_flag.is_enabled(course_key):
+            forced_into_bucket = (
+                bucket_flag.is_enabled(course_key) if course_key
+                else bucket_flag.is_enabled_without_course_context()
+            )
+            if forced_into_bucket:
                 bucket = i
                 break
+        else:
```

Comment:
Is this an `else` on a `for` loop? O_o

Reply:
Yup! Cute (IMO) Python feature that executes the `else` block iff the loop is not prematurely broken out of.

Reply:
Fun fact, Python also let's you use `else` along with a `try`/`except` block [1];
an `else` block is run if no exception if thrown in the `try` block.

- [1] https://docs.python.org/3/tutorial/errors.html